feat(codecs): Add syslog encoder

polarathene opened the pull request:
This PR has been opened at the request of others for providing feedback on the general direction and approach.

**NOTE:** The branch does not yet reflect the current progress. I had been working on a refactor of some changes earlier this year that ran into a blocker to completion for committing. I intended to resolve that first, but have not yet been able to afford the time.

---

I'll try to allocate time within October to get those changes applied to this branch, along with a proper PR description at that point.

polarathene commented:
> Any updates on this ?

No sorry. Several unexpected events happened in a short time period, I was unsuccessful at avoiding burnout which I'm still in the process of exiting from.

## Context

FWIW, I over committed myself in 2023, while facing several setbacks and issues in 2024. This isn't the only work where I'm pinged for updates to move things forward that have come to a halt.

I do understand the frustration of wanting to have this become actively developed again and merged, but without any assistance it's a task I struggle to allocate time for. More on that below.

---

### PR was opened due to demand

When I raised a concern about the lack of feedback received thus far, it was insisted that I open a PR, even if it's not in the state desired for handing to reviewers. 

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2348906692

> _Efforts are appreciated but AFAIK it is not equivalent to a PR; I'm pretty certain you are only able to comment and provide input with a pull request._

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2354138209

> _in order for reviewers to leave proper feedback for a branch it needs to have an issue or PR ticket. This would allow reviewers to make comparison between branches and might proceed with the better one._

However opening the PR early has made no difference vs the diff link I had previously provided. Which was anticipated:

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2334900866

> _When you say that we have nothing and making no progress without a PR, are you offering to participate in review?_
> _So far no one has provided any feedback on the linked branch/diff, so if a new PR left in a WIP status is opened will anything change?_

---

### My previous input

None of this has changed.
- I've been seeking feedback/guidance since July but none received yet, even after opening the PR as requested.
- I still want to work on it, but it is difficult to prioritize in my current situation.

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2354171857

> _I was hoping to just get general feedback if it was going in the right direction or if there was anything that stood out at a glance._

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2331625301

> _It would probably help to know how you'd like to use/configure it._
> 
> * _How important is it to have config for setting default/fallback values?_
> * _How important is it to have customization for keys vs managing it all via VRL?_
> 
> _Feedback on the last shared iteration would also be helpful to know if that's going in the right direction or not. I might have more time spare by October or November to finish up my branch._

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2357250235

> _I'd love to finish the PR sooner but have no employer at present, thus it's tricky to afford dedicating time to this over other priorities_ 😓

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2334900866

> _My work on it is purely voluntary, no employer backing, so dedicating my spare time on it is tricky_ 😓

https://github.com/vectordotdev/vector/pull/17668#issuecomment-2238655230

> _I do want to complete the work, it's just been difficult to budget my volunteer time for._
> _So if nobody helps, I will eventually be able to block time out again for coming back to this. I sunk a considerable chunk of time into the branch already, would be an awful waste to leave it to rot._

---

### Personal friction

When it comes to finding time to allocate to this PR, there's a bit more to it than covered below, but it's effectively a high cost item for me to tackle.

I need to have:
- No other events with higher priorities to resolve, or that unintentionally become time sinks (ADHD related)
- Sufficient time to context switch to this PR / task, and the energy to figure out and resolve any blocker or how to move forward (_receiving the feedback and guidance requested would be helpful_).
- Sufficient hardware resources. Often I am lacking enough memory on my system to confidently work on this PR. I juggle quite a lot and experience crashes where I lose a fair amount of work (_I do make an effort to save and add notes, but it's a tad complicated_). I also have very low storage spare on the system which has been a point of friction in the past for this PR.

As such, the most optimal time is when I have nothing pressing elsewhere and I'm able to save everything I can to perform a system reboot (_I only have a few of these a year as preparing for one can take multiple days for me_). I'd much rather have projects properly isolated to individual VMs which would address quite a lot of these problems and allow me to work much more efficiently, but that requires time.

If I had a separate system temporarily dedicated for working on this PR I could tackle it much more easily. Presently my goals are:
- Get a new phone (the one I've used since 2018 broke a few days ago)
- Finish up a documentation PR
- Prepare for system reboot and hope my system boots (_sometimes it appears bricked, or it fails in other ways_).
- Block out time for Vector and hope nothing interrupts getting to the sync point to update this PR branch.

Optimistically the system reboot will be completed by the 17th if nothing else interferes, so there should be a new push on this branch before the end of the month.

If that happens, I'd toggle of the draft status and ping for review/feedback. If I receive feedback in a timely manner progress will be quick from then on, but if there's a notable delay (over a week) I'd likely have switched to something else and it could take a while to return back to the PR. I'm slightly concerned about timing as we approach December, but hopefully we can get this PR back on track 💪

jcantrill commented:
@StephenWakely you commented in the previous PR, is it possible for you to provide additional feedback?  My limited familiarity with RUST makes my input less then useful

StephenWakely commented:
Sorry, I had assumed that since the PR is in draft you weren't looking for feedback yet, but I can definitely take a look

polarathene commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+    }
+}
+
+/// Syslog serializer options.
+#[configurable_component]
+#[derive(Clone, Debug, Default)]
+// Serde default makes all config keys optional.
+// Each field assigns either a fixed value, or field name (lookup field key to retrieve dynamic value per `LogEvent`).
+#[serde(default)]
+pub struct SyslogSerializerOptions {
+    /// RFC
+    rfc: SyslogRFC,
+    /// Facility
+    #[serde(deserialize_with = "Facility::deserialize")]
+    facility: Facility,
+    /// Severity
+    #[serde(deserialize_with = "Severity::deserialize")]
+    severity: Severity,
+
+    /// App Name
+    app_name: Option<String>,
+    /// Proc ID
+    proc_id: Option<String>,
+    /// Msg ID
+    msg_id: Option<String>,
+
+    /// Payload key
+    payload_key: String,
+
+    // Q: The majority of the fields above pragmatically only make sense as config for keys to query?
+}
```
This was adapted from the original PR, `payload_key` has since been removed in favor of the actual Vector log `message` when available or could be set via VRL.

In my local branch I was looking into having most of these field that have an equivalent syslog struct field to instead share the same struct:

```rust
pub struct SyslogSerializerOptions {
    /// RFC
    rfc: SyslogRFC,

    #[serde(flatten)]
    #[configurable(derived)]
    priority: Pri,

    #[serde(flatten)]
    #[configurable(derived)]
    tag: Tag,
}
```

There was another variation that instead of configuring overrides to the defaults, would use `ConfigTargetPath` as a more appropriate config approach by allowing the user to adjust the VRL keys to map into the syslog struct fields:

```rust
pub struct SyslogSerializerOptions {
    /// Facility
    #[default = "facility"]
    facility: ConfigTargetPath,
    /// Severity
    #[default = "severity"]
    severity: ConfigTargetPath,

    /// Application Name
    #[default = "appname"]
    app_name: ConfigTargetPath,
    /// Process ID
    #[default = "procid"]
    proc_id: ConfigTargetPath,
    /// Message ID
    #[default = "msgid"]
    msg_id: ConfigTargetPath,

    /// Hostname
    #[default = "hostname"]
    hostname: ConfigTargetPath,
    /// Message
    message: Option<ConfigTargetPath>,
    /// Timestamp
    timestamp: Option<ConfigTargetPath>,
    /// Structured Data
    structured_data: Option<ConfigTargetPath>,
}
```

Probably the right way to go?

polarathene reviewed (commented):
This is some parts I can recall that I would like to receive feedback on, I'll bring up other concerns next week.

StephenWakely commented on lib/codecs/Cargo.toml:
```diff
@@ -10,11 +10,13 @@ name = "generate-avro-fixtures"
 path = "tests/bin/generate-avro-fixtures.rs"
 
 [dependencies]
+akin = { version = "0.4", optional = true }
```
Generally if it is needed and the license is ok then we are fine to pull it in like you have done here.

However, I don't really see that this crate is adding enough to warrant it's inclusion. It's only used in one place from what I can see. There it would be more idiomatic to just use a simple macro:


```rust
macro_rules! deserialize_impl {
    ($enum:ty) => {
	impl $enum {
	    fn deserialize<'de, D>(deserializer: D) -> Result<Self, D::Error>
	
...

}

deserialize_impl!(Facility);
deserialize_impl!(Severity);
```

StephenWakely commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+    //
+    // Q: Why `$.message.` as the prefix? (Appears to be JSONPath syntax?)
+    // NOTE: Originally named in PR as: `get_field_or_config()`
+    fn replace_if_proxied(&self, value: &str) -> Option<String> {
```
I don't understand what this is doing, but whatever it is doing it is not what we want it to be doing... These function can just be removed given that the fields should be retrieved with something like

```
field = value.unwrap_or_else(|| log.get_field_by_meaning("service"))
```

With a different meaning for each field...

polarathene replied:
The original PR IIRC had implemented a feature to support config that overrides a default value (such as always X severity), presumably because the log event did not provide an equivalent mapping. Or the config key would reference a field to lookup and get the value of from the event, which I think their `$.message` prefix was specific to their own use-case.

I'd rather drop it, but wasn't sure how much of the original PR was the correct approach as I'm not familiar enough with Vector myself. Thus this PR focused on refactoring the functionality that was already in the existing PR and moving forward from there 😅 

---

So this method was mostly a helper:

```rust
        value
            .strip_prefix("$.message.")
            .map_or(
                Some(value.to_owned()),
                |field_key| self.value_by_key(field_key),
            )
    }

    fn value_by_key(&self, field_key: &str) -> Option<String> {
        self.log.get(field_key).map(|field_value| {
            field_value.to_string_lossy().to_string()
        })
    }
```

Used earlier like:

```rust
    fn decant_config(&self, config: &SyslogSerializerOptions) -> SyslogMessage {
        let y = |v| self.replace_if_proxied_opt(v);
        let app_name = y(&config.app_name).unwrap_or("vector".to_owned());
        let proc_id = y(&config.proc_id);
        let msg_id = y(&config.msg_id);
```

Since they're options, if no lookup key or default value was configured, `y` would return `None`:
- `proc_id` and `msg_id` would later be formatted as `-`.
- `app_name` default is meant to fallback to something IIRC so `vector` was the default chosen. In my local branch I used the `SmartDefault` crate to add a default to the actual syslog `Tag` struct for the `app_name` field instead.

---

I'm not familiar with `log.get_field_by_meaning` or if that was available when I started working on this PR early this year. I'll look into that if it's appropriate (_I'm aware of [this reference](https://github.com/vectordotdev/vector/pull/17668#discussion_r1761846388) you shared in another feedback comment, I haven't gone over that yet_).

IIRC this logic may have been refactored on my local branch.

jcantrill replied:
> The original PR IIRC had implemented a feature to support config that overrides a default value (such as always X severity), presumably because the log event did not provide an equivalent mapping. Or the config key would reference a field to lookup and get the value of from the event, which I think their $.message prefix was specific to their own use-case.

This came out of translating functionality from an existing fluentd plugin.  One can accomplish similar need by using existing vector functionality to template config field and walk a path.  IMO you can remove this functionality if it is possible to do a similar way in vector already

polarathene replied:
> IMO you can remove this functionality if it is possible to do a similar way in vector already

Pretty sure I used VRL with a remap transform in my vector config to get similar functionality. I'm not fond of using config settings for the encoder to support individual fields that could be either a fixed value or lookup key.

When I've had time to refresh my familiarity with the PR source and sync my local changes to this branch we can discuss further.

Either way it'll need to clearly document how a user can map their source log event into the expected input/format that the encoder needs, relying on config settings for this was mostly a convenience IIRC but introduced some added complexity that I don't think was common with other official Vector encoders.

StephenWakely commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+    }
+
+    fn decant_config(&self, config: &SyslogSerializerOptions) -> SyslogMessage {
+        let y = |v| self.replace_if_proxied_opt(v);
+        let app_name = y(&config.app_name).unwrap_or("vector".to_owned());
+        let proc_id = y(&config.proc_id);
+        let msg_id = y(&config.msg_id);
```
These should all be set like:

```rust
field = config.value.unwrap_or_else(|| log.get_field_by_meaning("service"))
```

StephenWakely commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+        // https://github.com/vectordotdev/vector/blob/ad6a48efc0f79b2c18a5c1394e5d8603fdfd1bab/lib/vector-core/src/event/log_event.rs#L532-L541
+        // `v.to_string_lossy()`:
+        // https://github.com/vectordotdev/vrl/blob/f2d71cd26cb8270230f531945d7dee4929235905/src/value/value/serde.rs#L34-L55
+        let payload = if config.payload_key.is_empty() {
+            self.log.get_message().map(|v| v.to_string_lossy().to_string() )
+        } else {
+            self.value_by_key(&config.payload_key)
+        };
+
+        payload.unwrap_or_default()
```
No payload key, so just something like this should do:

```suggestion
            self.log.get_message().map(|v| v.to_string_lossy().to_string())
```

StephenWakely commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+                // https://datatracker.ietf.org/doc/html/rfc3164#section-4.1.2
+                // https://docs.rs/chrono/latest/chrono/format/strftime/index.html
+                //
+                // TODO: Should this remain as UTC or adjust to the local TZ of the environment (or Vector config)?
```
Keep it UTC.

StephenWakely commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+            &self.message,
+        ].concat()
+
+        // Q: RFC 5424 MSG part should technically ensure UTF-8 message begins with BOM?
```
If you want the challenge you are welcome to embark on the journey to make this comply. However, I feel that this occurs in practice rarely enough that it won't be a blocker to getting this merged.

StephenWakely commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+    }
+
+    fn get_structured_data(&self) -> Option<StructuredData> {
+        self.log.get("structured_data")
```
For the Vector namespace, (when `log_namespace: true`) we want to get this by meaning instead of name:

```suggestion
  match self.log.namespace() {
       LogNamespace::Vector => self.log.get_by_meaning("structured_data")
       LogNamespace::Legacy => self.log.get("structured_data")
  }
```

StephenWakely reviewed (changes requested):
Thanks for your efforts in putting this together! I appreciate it has been a bit of journey.

Mostly the changes needed are to make it work better when we have `log_namespace: true` set. 

There's a few other checks that need to pass in order to merge this. See [here](https://github.com/vectordotdev/vector/blob/master/CONTRIBUTING.md#running-other-checks) for more details.

Any questions, please give me a shout.

polarathene reviewed (commented):
Thanks for the speedy feedback! ❤️ 

I'll return to this next week to go over your feedback properly then and address it, cheers 😁

jcantrill commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+    msg_id: Option<String>,
+
+    /// Payload key
+    payload_key: String,
```
| Remove this option. This should always be the field with the meaning message.

I would disagree with this suggestion as it reduces the flexibility of using the protocol.  This implies you always want the "message" portion of the payload to be what vector determines is the "message" and not for instance how an administrator may have transformed it in a prior pipeline to be something else.  The default if not specified IMO should be "message" but allow the flexibility; there is no reason not to

polarathene replied:
> This implies you always want the "message" portion of the payload to be what vector determines is the "message" and not for instance how an administrator may have transformed it in a prior pipeline to be something else.

Could you elaborate why you'd have introduced a difference in semantics? Please document a valid use-case for why you'd want such functionality so we know it's not an XY problem.

When would you transform the log message to not represent the message, yet provide it as input to an encoder to but require a different key for the actual message content?

---

I am familiar with the original PR use with `add_source` to compose metadata related to k8s with the actual message. You could do that with VRL, or for the metadata leverage structured data?

jcantrill replied:
> Could you elaborate why you'd have introduced a difference in semantics? Please document a valid use-case for why you'd want such functionality so we know it's not an XY problem.

Vector is configurable as a transformation engine.  Logs coming in do not need to equal logs going out.  Consider almost any  case where my desired payload is a subset of the original message, a value in the message which was a JSON string.  In a prior transform, I could parse the message and now it is available to me as key/value pairs, where I only care about some portion of the message.

By removing this option does it mean in some prior transform do I always have to modify the event and set the key "message" to the desired outcome?

polarathene replied:
> By removing this option does it mean in some prior transform do I always have to modify the event and set the key "message" to the desired outcome? Maybe I'm missing where the encoding ends and transforms and sinks begin 🤷‍♂️

Encoding is `A => B`, given input A produce output B (syslog formatted log in this case).

Perhaps I should phrase this a bit differently?:
- If Vector supports encoding your input to other existing log encoders the way you want, you should expect to get the same with this syslog encoder.
- If it does not, then there would need to be a very valid reason why this encoder is an exception vs what you'd otherwise need to do with any other encoder already supported by Vector.

As such, you should find that what you're requesting isn't specific to this encoder implemented via the PR, but something one would expect to be encoder agnostic.

That would be a separate feature request to raise, which may be rejected on the basis that VRL would be advised for your use-case to pre-process your input for the encoder instead.

---

I'm not an active Vector user myself yet, but from what I do understand `message` is an internal convention and the encoders rely on that to be semantically correct as input, thus you'd be expected to use existing Vector features to handle that prior to encoding.

@StephenWakely will chime in if I'm mistaken, and you're also welcome to correct me too.

jcantrill commented on lib/codecs/src/encoding/format/syslog.rs:
```diff
@@ -0,0 +1,577 @@
+    /// Proc ID
+    proc_id: Option<String>,
+    /// Msg ID
+    msg_id: Option<String>,
```
> This option is not needed.
> If the message contains a field with a meaning of proc_id, we should use that field. Otherwise, this field should be the PID of Vector.

Quoting the spec:
```   
PROCID is a value that is included in the message, having no
   interoperable meaning, except that a change in the value indicates
   there has been a discontinuity in syslog reporting.  The field does
   not have any specific syntax or semantics; the value is
   implementation-dependent and/or operator-assigned.  The NILVALUE MAY
   be used when no value is provided.
   ```
We use vector in a containerized environment so this being something like say... the podname, or pod uid is more useful then pid... pid is always 1 in the container.  Being able to set this from unique pod information is useful

> Can you remove this option and take the value from the event with a meaning of msgid. We should also update the decoder to insert msgid with this meaning.

Again, is there a reason not to provide a default and make this configurable.   The spec says: 
```
 The MSGID SHOULD identify the type of message.  For example, a
   firewall might use the MSGID "TCPIN" for incoming TCP traffic and the
   MSGID "TCPOUT" for outgoing TCP traffic. Messages with the same
   MSGID should reflect events of the same semantics.  The MSGID itself
   is a string without further semantics.  It is intended for filtering
   messages on a relay or collector.
```
which means to me there may be a possible for a default but it also could be something an admin wants to configure.  We use vector in to kubernetes cluster from various log sources.  I can imagine using this field to distinguish between container and node journal logs

jcantrill commented:
@StephenWakely @polarathene 

I don't fully understand all of what [this](https://github.com/vectordotdev/vector/pull/21307#discussion_r1842067705) entails but my general impression is the suggested changes are moving from flexible to an opinionated implementation, relying upon the impl to assume values.  Maybe I'm reading incorrectly but I am pushing to ask for flexibility and configurability.  The spec defines specific attributes and layout of a message but leaves quite a bit open for interpretation by an admin.   As a user of vector in a cloud containerized environment that already is forwarding logs over syslog, we have use cases where it makes sense to allow administrators to set various fields.  

The original PR as a straight port of a fluend plugin and what we have now for our vector implementation. Reviewing the syslog spec I can see where several config knobs should be dropped (e.g. tag, add_source) but IMO the others should remain

polarathene commented:
> my general impression is the suggested changes are moving from flexible to an opinionated implementation, relying upon the impl to assume values.
> Maybe I'm reading incorrectly but I am pushing to ask for flexibility and configurability.

I haven't had time to go over full review feedback, but I'll do so next week when I return to sync my local changes to this branch.

- Any input from me thus far in the review process is mostly focused on the encoder config settings which IMO were not implemented well in the original PR.
- My primary focus will be on getting the features core functionality sorted for review and approval. At which point I'd reach out for subscribed users to test it out for feedback so that we can verify the feature is useful your use-cases.

Config wise, I think any property/field mapping can be handled via VRL with remap. We can discuss that in future if it's inconvenient and could be improved, but given my limited time I hope you understand why I don't want too much bike shedding over that right now.

----

> The spec defines specific attributes and layout of a message but leaves quite a bit open for interpretation by an admin.
> As a user of vector in a cloud containerized environment that already is forwarding logs over syslog, we have use cases where it makes sense to allow administrators to set various fields.

When this PR reaches the next milestone (when I ping for proper review / testing), I'd be happy to discuss this with you here. As per my review on the original PR, it should be evident that I care about respecting the spec. Anything that should be `-` (NIL) when not present is handled, and you should have the ability to override each field IIRC, just via Vector remap transforms.

FWIW: My interest in Vector with this feature is also for usage in a container.

---

> The original PR is a straight port of a fluentd plugin and what we have now for our vector implementation.

This PR adapted the original, you can follow the commit history for logical commits with descriptions that reason each change.

IIRC the original had various concerns that I've addressed or intend to (_these should be detailed in the commit messages and/or inline commentary_).

I remember the `add_source` functionality and dropping it here in one of the commits as it seemed to be redundant (could be handled via VRL), thus I think most of your concerns aren't likely to be a problem and you'd still have the control/flexibility for your requirements, it just may look a little different config wise.

StephenWakely commented:
> @StephenWakely @polarathene
> 
> I don't fully understand all of what [this](https://github.com/vectordotdev/vector/pull/21307#discussion_r1842067705) entails but my general impression is the suggested changes are moving from flexible to an opinionated implementation, relying upon the impl to assume values. Maybe I'm reading incorrectly but I am pushing to ask for flexibility and configurability. 

It's pushing the flexibility back into VRL, which we already have. This codec should be opinionated about what it expects the log message to look like. Any flexibility you need should be handled in a remap transform.

pront commented:
> > @StephenWakely @polarathene
> > I don't fully understand all of what [this](https://github.com/vectordotdev/vector/pull/21307#discussion_r1842067705) entails but my general impression is the suggested changes are moving from flexible to an opinionated implementation, relying upon the impl to assume values. Maybe I'm reading incorrectly but I am pushing to ask for flexibility and configurability.
> 
> It's pushing the flexibility back into VRL, which we already have. This codec should be opinionated about what it expects the log message to look like. Any flexibility you need should be handled in a remap transform.

This sounds like a good strategy in general. Remap can do any kind of event transformation you need. The encoder should specify the interface it expects so it can efficiently batch and send requests downstream.

jcantrill commented:
@polarathene any progress on driving this forward?  
@StephenWakely @polarathene anything we can do to accelerate the adoption of this PR?  Is there a place to land something and continue with any gaps?

pront commented:
> @polarathene any progress on driving this forward? @StephenWakely @polarathene anything we can do to accelerate the adoption of this PR? Is there a place to land something and continue with any gaps?

We can introduce it as a beta feature. Before merging though I would like to see all the open comments addressed. Feel free to collaborate and push directly to this PR if possible. When in doubt, decide based on this https://github.com/vectordotdev/vector/pull/21307#issuecomment-2483009183

polarathene commented:
> @polarathene any progress on driving this forward?

Sorry life and other commitments has made it hard to prioritize time for this. I still want to see it through, the earliest next window I'd have for that is early May.

I'll chime in again if I resume work on this PR. The least I can do is at least get the remaining changes I made sync'd back to this PR so that others can continue it if I'm unable to find the time to wrap it up. Apologies for how long this has taken.

jcantrill commented:
> The least I can do is at least get the remaining changes I made sync'd back to this PR so that others can continue it if I'm unable to find the time to wrap it up

@polarathene Have you made any progress towards :arrow_up: this comment?  I believe we may have resources that are available to push this PR forward but need a base from which they can work and maybe a list of what you believe still needs to be addressed.  Thank you for your efforts

BinarHugo commented:
Good afternoon, is there any information on continuing work on this topic?

polarathene commented:
@vparfonov ah that's great thank you! I began resuming work late august to rebase the PR and start addressing the review feedback but injured myself an I'm still recovering. I have little time that I can use at my PC.

I will try to allocate some time to look over your PR in the next day or so depending how well I am (_my back and neck muscles cannot take much stress presently as I overexerted them so I am mostly confined to bed at present_).

I did notice at a glance over your PR description that you (understandably) made changes that @jcantrill was seeking rather than what review feedback of this PR requested instead as per @pront and @StephenWakely 

It may be better to stash out the extra customizations into a follow-up PR/issue as they don't sound like a feature that would be specific to a syslog encoder (_shouldn't the remap functionality already address some of the custom support that your PR wants to introduce, there is also the [semantic meaning keys](https://github.com/vectordotdev/vector/pull/21307#discussion_r1842067705) for config to leverage_)

---

If your PR is reviewed and merged that would be great, otherwise once I get the all clear to return to working at my desk again I'll return to resume this PR if desired.
